kingpin/actors/packagecloud.py

```diff
@@ -0,0 +1,256 @@
+            for package in packages:
+                self.log.debug('Inspecting package %s' % package)
+
+                if len(packages) >= self.option('number_to_keep'):
+                    if not by_date:
+                        continue
+                    else:
+                        if datetime.datetime.now() - package['created_at'] > \
+                           datetime.timedelta(
+                               seconds=self.option('older_than')):
+                            pass
+                        else:
+                            continue
+
+                    msg = '%s/%s/%s' % (
+                        self.option('repo'), package['distro_version'],
+                        package['filename'])
+                    if self._dry:
+                        log.info('Would have deleted %s' % msg)
+                    else:
+                        log.info('Deleting %s' % msg)
+
+                        # Decrement list of packages to track how many are left
+                        packages.remove(package)
+
+                        yield self._packagecloud_client.delete(
+                            token=TOKEN,
+                            account=ACCOUNT,
+                            repo=self.option('repo'),
+                            distro_version=package['distro_version'],
+                            filename=package['filename']
+                        ).http_delete()
+
+
+class Delete(PackagecloudBase):
+
+    """packagecloude Delete Base Object"""
+
+    all_options = {
+        'number_to_keep': (
+            int, 0,
+            'Keep at least this number of each package'),
+        'packages_to_delete': (
+            str, REQUIRED,
+            'Regex of packages to delete, e.g. package1|package2'),
+        'repo': (
+            str, REQUIRED,
+            'Which packagecloud repo to delete from'),
+    }
+
+    @gen.coroutine
+    def _execute(self):
+        """Deletes all packages that match the `packages_to_delete` pattern
+
+        raises: gen.Return()
+        """
+        yield self._delete()
+
+
+class DeleteByDate(PackagecloudBase):
+
+    """packagecloude DeleteByDate Base Object"""
+
+    all_options = {
+        'number_to_keep': (
+            int, 0,
+            'Keep at least this number of each package'),
+        'older_than': (
+            int, REQUIRED,
+            'Delete packages created before this number of seconds'),
+        'packages_to_delete': (
+            str, REQUIRED,
+            'Regex of packages to delete, e.g. package1|package2'),
+        'repo': (str, REQUIRED,
+                 'Which packagecloud repo to delete from')
+    }
+
+    @gen.coroutine
+    def _execute(self):
+        """Deletes all packages that match the `packages_to_delete` pattern and
+        where created before `older_than` number of seconds
+
+        raises: gen.Return()
+        """
+        yield self._delete(by_date=True)
```

Comment:
The general pattern we've been following is to make the `execute()` methods have the raw _logic_ for the actor, and then just build supporting private methods that can be re-used and well tested. So for example, you may have the following private methods:

```
def _find_package(pattern, repo):
  """Returns a list of packages (with metadata about each one) that matches *pattern*"""
```

```
def _find_by_date(pattern, repo, age=5, minimum=1):
  """Finds a list of packages that match *pattern*, and are older than *age* days old, while keeping *minimum* no matter what."""
  packages = yield self._find_packages(pattern)
  for package in packages:
    ... do stuff
  raise gen.Return(deletable_packages)
```

```
def _delete_package(name, repo, version=None):
  """Deletes a package that matches *version*, or all packages if no *version* is supplied."""
```

Then you're `execute()` method looks something like:
```
def _execute(self):
  to_delete = yield self._find_by_date(options.name, options.older_than, options.number_to)keep)
  if self._dry:
    log something..
  yield self._delete(to_delete)
```